---
output: github_document
---
<!-- README.md is generated from README.Rmd. Please edit README.Rmd file -->

# `MiXcan: Statistical Framework for Cell-type-Aware Transcriptome-Wide Association Studies with Bulk Tissue Data`

## Introduction to **MiXcan**

**Goal:** 

* Constructs cell-type-level prediction models for genetically regulated expression (GReX);

* Predicts cell-type-level GReX in new genotype data; and

* Performs cell-type-aware TWAS.

**Advantages over tissue-level TWAS:**

* Improves GReX prediction accuracy; 

* Boosts the study power, especially for genes that function in minor cell types or have different association directions in different cell types;

* Sheds light on the responsible cell type(s) of  associations. 

**Disadvantages over tissue-level TWAS:**

* Requires prior knowledge on disease-critical cell types and their proportions in tissue;

* Has more model parameters;

* May be less powerful than tissue-level TWAS for genes that have similar disease associations in different cell types or function in major cell types. 


**Input:**

* Prediction model construction: genotype, covariates, and gene expression data (same as in PrediXcan) + cell-type composition estimates (e.g. from existing methods, such as ESTIMATE, CIBERSORT, xCell). 

* Association Analysis: genotype, covariates and phenotype data (same as in PrediXcan).


**Output:**

* Prediction model construction: Cell-type-specific or nonspecific prediction weights for different genes. 

* Association Analysis: Tissue-level association p-values and cell-type-level association summaries including estimates, standard error and p-values.




A full description of the method can be found in our [paper](https://www.biorxiv.org/content/10.1101/2022.03.15.484509v1.abstract).

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE)
library(MiXcan2) 
```


## Installation

### Hardware Requirements

The MiXcan package requires only a standard computer with a reasonable RAM/CPU to support the operations. The minimal RAM recommended by authors is 2 GB. Authors used a computer with the following specification:

RAM: 32 GB

CPU: 2.3 GHz 8-core

### Software Requirements

The github package is developed on macOS operating systems. The MiXcan pacakge is developed under an open source software R (version 4.1.2). Different versions of R can be downloaded at https://cran.r-project.org/. 

### Package Installation

With R, users can install the MiXcan package directly from GitHub with [devtools](https://github.com/hadley/devtools):

```R
install.packages("devtools")
devtools::install_github("songxiaoyu/MiXcan2/Package")

````
The typical install time of the package is less than 5 minutes. 

## Example of use

Below demonstrates the MiXcan analysis pipeline on a single peusdo gene. In reality, multiple genes can be analyzed in parallel. Holdout set can be pre-excluded to allow model training on the remaining samples. 

### Data

Create peusdo data for demonstration:

```{r}
set.seed(123)
n=200
pi=rbeta(n, 2, 3)
x=matrix(rbinom(n*10, 2, 0.3), ncol=10)
y=x[,1]*pi+rnorm(n, sd=0.2)
cov=NULL
```

### MiXcan analysis pipeline
```{r}
library(doParallel)
library(tidyverse)
library(rlist)
nCores=detectCores()-1; registerDoParallel(nCores) # use parallel computing for speed, but leave 1 core out for other activities. 

```

Step 2: Estimating cell-type-specific (and nonspecific) GReX prediction weights of a gene using the MiXcan function

```{r}
set.seed(111)
foldid <- sample(1:10, length(y), replace=T)

model <- MiXcan2_model(y=y, x=x, cov = cov,
                        pi= pi,
                        foldid = foldid, yName="Gene1",
                       xNameMatrix = paste0("X", 1:10))
model$beta.SNP.cell1
model$beta.SNP.cell2
```

Step 3: Extracting the weights and model summaries from the MiXcan output.
```{r}
MiXcan_weight_result <- MiXcan2_extract_weight(model = model)
MiXcan_weight_result

```

```{r}
MiXcan_summary_result <- MiXcan2_extract_summary(x=x, y=y, cov=cov,
                                                pi=pi, model=model)
MiXcan_summary_result
```
Note, the MiXcan estimated models are from penalized regression (elastic-net), which shrinks the effect size towards zero. If users are interested to use less penalized weights for the MiXcan selected SNPs, they can employ the following function: 
```{r}
MiXcan_refit <- MiXcan2_refit(model = model,
                                           y=y,
                                           x=x, cov = cov,
                                           pi= pi)
MiXcan_refit$weight
MiXcan_refit$summary
```
One can run MiXcan for multiple times to ensemble models for downstream anlaysis. Here are the codes:
```{r}
ensemble=MiXcan2_ensemble(y=y, x=x, cov=cov, pi=pi, 
                 yName="Gene1", B=10, seed=123) 
  
ensemble$summary
ensemble$ensemble_weight
ensemble$all_weights
```

